nifi-extension-bundles/nifi-kafka-bundle/nifi-kafka-processors/src/test/java/org/apache/nifi/kafka/processors/producer/convert/KafkaRecordConverterTest.java

```diff
@@ -0,0 +1,155 @@
+
+@ExtendWith(MockitoExtension.class)
+class KafkaRecordConverterTest {
+
+    private static final int EXPECTED_RECORD_COUNT = 50000; // Expect 50,000 records
+    private static final int MAX_MESSAGE_SIZE = 5 * 1024 * 1024; // 5MB
+    private static final byte[] LARGE_SAMPLE_INPUT = new byte[MAX_MESSAGE_SIZE];
+    private static final Map<String, String> SAMPLE_ATTRIBUTES = Map.of("attribute1", "value1");
+
+    @Mock
+    private HeadersFactory headersFactory;
+
+    @Mock
+    private KeyFactory keyFactory;
+
+    private JsonTreeReader jsonTreeReader;
+
+    private JsonRecordSetWriter jsonRecordSetWriter;
+
+    @BeforeEach
+    void setUp() throws Exception {
+        TestRunner runner = TestRunners.newTestRunner(new PublishKafka());
+
+        MockitoAnnotations.openMocks(this);
+
+        jsonTreeReader = new JsonTreeReader();
+
+        runner.addControllerService("jsonTreeReader", jsonTreeReader);
+        runner.setProperty(jsonTreeReader, SchemaAccessUtils.SCHEMA_ACCESS_STRATEGY, "infer-schema");
+        runner.enableControllerService(jsonTreeReader);
+
+        jsonRecordSetWriter = new JsonRecordSetWriter();
+
+        runner.addControllerService("jsonRecordSetWriter", jsonRecordSetWriter);
+        runner.setProperty(jsonRecordSetWriter, SchemaAccessUtils.SCHEMA_ACCESS_STRATEGY, SchemaAccessUtils.INHERIT_RECORD_SCHEMA);
+        runner.setProperty(jsonRecordSetWriter, "output-grouping", "output-oneline");
+        runner.enableControllerService(jsonRecordSetWriter);
+
+        populateSampleInput();
+    }
+
+    // Create 5MB of sample data with multiple records
+    private static void populateSampleInput() {
+        StringBuilder sb = new StringBuilder();
+        int recordCount = EXPECTED_RECORD_COUNT;
+        int approximateRecordSize = MAX_MESSAGE_SIZE / recordCount;
+
+        for (int i = 0; i < recordCount; i++) {
+            sb.append("{\"key\": \"").append(i).append("\",\"value\":\"");
+            int payloadSize = approximateRecordSize - 30;
+            for (int j = 0; j < payloadSize; j++) {
+                sb.append((char) ('A' + (j % 26)));
+            }
+            sb.append("\"}");
+            if (i < recordCount - 1) {
+                sb.append("\n");
+            }
+            if (sb.length() >= MAX_MESSAGE_SIZE - 100) {
+                break;
+            }
+        }
+
+        byte[] stringBytes = sb.toString().getBytes();
+
+        Arrays.fill(LARGE_SAMPLE_INPUT, (byte) ' ');  // Fill with spaces instead of NULL
+
+        int copyLength = Math.min(stringBytes.length, LARGE_SAMPLE_INPUT.length);
+        System.arraycopy(stringBytes, 0, LARGE_SAMPLE_INPUT, 0, copyLength);
+    }
+
+    private void verifyRecordCount(Iterator<KafkaRecord> records, int expectedCount) {
+        List<KafkaRecord> recordList = new ArrayList<>();
+        records.forEachRemaining(recordList::add);
+        assertEquals(expectedCount, recordList.size(), "Expected exactly " + expectedCount + " records");
+    }
```

Comment:
This method and the following 4 tests should all be part of a `ParamaterizedTest`